specifyweb/backend/trees/extras.py

```diff
@@ -742,3 +742,80 @@ def is_treedef(obj):
     return issubclass(obj.__class__, Tree) or bool(
         re.search(r"treedef'>$", str(obj.__class__), re.IGNORECASE)
     )
+
+def renumber_tree(table: str) -> None:
+    cursor = connection.cursor()
+    logger.debug(f"[renumber_tree] running for {table}")
+
+    cursor.execute("SET SESSION sql_safe_updates = 0") # might be needed for MariaDB 11
+
+    # Sync rankid
+    sql_sync = (
+        f"UPDATE {table} t "
+        f"JOIN {table}treedefitem d ON t.{table}treedefitemid = d.{table}treedefitemid "
+        f"SET t.rankid = d.rankid"
+    )
+    logger.debug(sql_sync)
+    cursor.execute(sql_sync)
+
+    # Compute max logical depth to size the LEFT JOIN ladder
+    cursor.execute(f"SELECT COUNT(DISTINCT rankid) FROM {table}")
+    depth = cursor.fetchone()[0] or 1
+
+    def tree_parent_joins(tbl: str, d: int) -> str:
+        if d <= 1:
+            return ""  # only t0 exists
+        return "\n".join(
+            f"LEFT JOIN {tbl} t{j} ON t{j-1}.parentid = t{j}.{tbl}id"
+            for j in range(1, d)
+        )
+
+    def tree_path_expr(tbl: str, d: int) -> str:
+        # Highest ancestor first, leaf last; LPAD stabilizes lexical order
+        parts = ", ".join([f"LPAD(t{i}.{tbl}id, 12, '0')" for i in reversed(range(d))])
+        return f"CONCAT_WS(',', {parts})"
+
+    # Preorder numbering using ROW_NUMBER()
+    sql_preorder = (
+        f"UPDATE {table} t\n"
+        f"JOIN (\n"
+        f"  SELECT id, rn FROM (\n"
+        f"    SELECT\n"
+        f"      t0.{table}id AS id,\n"
+        f"      ROW_NUMBER() OVER (ORDER BY {tree_path_expr(table, depth)}, t0.{table}id) AS rn\n"
+        f"    FROM {table} t0\n"
+        f"{tree_parent_joins(table, depth)}\n"
+        f"  ) ordered\n"
+        f") r ON r.id = t.{table}id\n"
+        f"SET t.nodenumber = r.rn,\n"
+        f"    t.highestchildnodenumber = r.rn"
+    )
+    logger.debug(sql_preorder)
+    cursor.execute(sql_preorder)
+
+    # Compute highestchildnodenumber bottom-up
+    sql_ranks = f"SELECT DISTINCT rankid FROM {table} ORDER BY rankid DESC"
+    logger.debug(sql_ranks)
+    cursor.execute(sql_ranks)
+    ranks = [row[0] for row in cursor.fetchall()]
+
+    for rank in ranks[1:]:
+        sql_hcnn = (
+            f"UPDATE {table} t\n"
+            f"JOIN (\n"
+            f"  SELECT parentid, MAX(highestchildnodenumber) AS hcnn\n"
+            f"  FROM {table}\n"
+            f"  WHERE rankid > %(rank)s\n"
+            f"  GROUP BY parentid\n"
+            f") sub ON sub.parentid = t.{table}id\n"
+            f"SET t.highestchildnodenumber = sub.hcnn\n"
+            f"WHERE t.rankid = %(rank)s"
+        )
```

Comment:
(Just a note!)
I tested node numbering with different taxon trees in the same discipline

![Image](https://github.com/user-attachments/assets/c1f0de4c-ab2d-4a41-9aea-90377cdb21c7)